### PR TITLE
Invite user and use mailgun

### PR DESCRIPTION
**Pull request by gvszg**

### CODE REVIEW

File: app/controllers/invitations_controller.rb [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+  def create
+    @invitation = Invitation.new(invitation_params)
+
+    if @invitation.save
+      AppMailer.send_invitation_email(@invitation).deliver
+      flash[:success] = "You have successfully invited #{@invitation.recipient_name}!"
+      redirect_to new_invitation_path
+    else
+      flash.now[:danger] = "Please check all information!"
```
Comment (tomtomecek) [inline app/controllers/invitations_controller.rb]: Good job!

File: app/controllers/invitations_controller.rb [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+      render :new
+    end
+  end
+
+  private
+
+  def invitation_params
+    params.require(:invitation).permit(:recipient_name, :recipient_email, :message).merge!(inviter_id: current_user.id)
```
Comment (tomtomecek) [inline app/controllers/invitations_controller.rb]: I don't think you need to `inviter_id` - should work with just `inviter` and its shorter.

File: app/controllers/reset_passwords_controller.rb [hunk 3]
```diff
@@ -13,8 +13,8 @@ def create
     user = User.find_by(token: params[:token])
     if user
       user.password = params[:password]
-      user.generate_token
       user.save
+      user.clear_token
```
Comment (tomtomecek) [inline app/controllers/reset_passwords_controller.rb]: This method is a good job!

File: app/controllers/reset_passwords_controller.rb [hunk 4]
```diff
@@ -13,8 +13,8 @@ def create
     user = User.find_by(token: params[:token])
     if user
       user.password = params[:password]
-      user.generate_token
       user.save
```
Comment (tomtomecek) [inline app/controllers/reset_passwords_controller.rb]: setting password and saving is the same as using `update` method.

Comment (tomtomecek) [inline app/controllers/reset_passwords_controller.rb]: Also what happens when password is invalid?

File: app/controllers/users_controller.rb [hunk 5]
```diff
@@ -26,4 +38,13 @@ def show
   def user_params
     params.require(:user).permit(:email, :password, :username)
   end
+
+  def handle_invitation
+    if params[:invitation_token].present?
+      invitation = Invitation.find_by(token: params[:invitation_token])
+      @user.follow(invitation.inviter)
+      invitation.inviter.follow(@user)
+      invitation.update_column(:token, nil)
```
Comment (tomtomecek) [inline app/controllers/users_controller.rb]: This could use a refactor to Invitation model.

File: app/models/user.rb [hunk 6]
```diff
@@ -37,6 +39,10 @@ def total_number_of_people_who_follow_me
   end
 
   def generate_token
-    self.token = SecureRandom.urlsafe_base64
+    self.update_column(:token, SecureRandom.urlsafe_base64)
+  end
+
+  def clear_token
+    self.update_column(:token, nil)
```
Comment (tomtomecek) [inline app/models/user.rb]: There isn't a need to use `self` here.

File: app/models/user.rb [hunk 7]
```diff
@@ -37,6 +39,10 @@ def total_number_of_people_who_follow_me
   end
 
   def generate_token
-    self.token = SecureRandom.urlsafe_base64
+    self.update_column(:token, SecureRandom.urlsafe_base64)
```
Comment (tomtomecek) [inline app/models/user.rb]: There isn't a need to use `self` here.

File: app/views/invitations/new.html.haml [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+  .row
+    .col-sm-10.col-sm-offset-1
+      = bootstrap_form_for @invitation, class: 'invite' do |f|
+        %header
+          %h1 Invite a friend to join MyFlix!
+        %fieldset
+          .form-group
+            .row
+              .col-sm-4
+                = f.text_field :recipient_name, class: "form-control", label: "Friend's Name"
```
Comment (tomtomecek) [inline app/views/invitations/new.html.haml]: When you use `bootstrap_form_for` then `#text_field` method will provide `form-group` div with `row` and `col-sm-x`.

HOMEWORK: study bootstrap_form_for gem docs and check in your browser what it produces.

File: app/views/invitations/new.html.haml [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+          .form-group
+            .row
+              .col-sm-4
+                = f.email_field :recipient_email, class: "form-control", label: "Friend's Email Address"
```
Comment (tomtomecek) [inline app/views/invitations/new.html.haml]: same here as above.

File: app/views/invitations/new.html.haml [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+          .form-group
+            .row
+              .col-sm-6
+                = f.text_area :message, rows: 6, class: "form-control", placeholder: "Please join this really cool site!", label: "Invitation Message"
```
Comment (tomtomecek) [inline app/views/invitations/new.html.haml]: Same as above, refactor this please.

File: app/views/invitations/new.html.haml [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+        %fieldset.form-group.action
+          %input(type="submit" value="Send Invitation" class="btn btn-default")
```
Comment (tomtomecek) [inline app/views/invitations/new.html.haml]: You should use `f.submit ...` when you use Rails helpers above.

File: app/controllers/reset_passwords_controller.rb (hunk 4 above)
Comment (gvszg) [inline app/controllers/reset_passwords_controller.rb]: When password invalid showing this:
```ruby
  flash[:danger] = "Password cannot be blank!"
  redirect_to "/reset_passwords/#{user.token}"
```

Is that proper?

File: app/views/invitations/new.html.haml (hunk 8 above)
Comment (gvszg) [inline app/views/invitations/new.html.haml]: I refactored this template.
The column length actually showed `col-sm-10` not `col-sm-4` like before.
I found the form with `layout: :horizontal` can use `control_col: "col-sm-4"`.
But it looks not good.

File: app/controllers/users_controller.rb (hunk 5 above)
Comment (gvszg) [inline app/controllers/users_controller.rb]: I did it this way:

in `app/controllers/user_controller.rb`
```ruby 
private

def handle_invitation
  if params[:invitation_token].present?
    invitation = Invitation.find_by(token: params[:invitation_token]) 
    @user.follow(invitation.inviter)
    invitation.inviter_follow_user(@user)    
  end   
end
```

in `app/models/invitation.rb`
```ruby
def inviter_follow_user(user)
  inviter.follow(user)
  update_column(:token, nil)
end
```

Are there better ways to refactor?

File: app/controllers/reset_passwords_controller.rb (hunk 4 above)
Comment (tomtomecek) [inline app/controllers/reset_passwords_controller.rb]: You should not hardcode the path

File: app/controllers/users_controller.rb (hunk 5 above)
Comment (tomtomecek) [inline app/controllers/users_controller.rb]: thats fine, but i thought you would refactor this `update_column(:token, nil)` only.

File: app/views/invitations/new.html.haml (hunk 8 above)
Comment (tomtomecek) [inline app/views/invitations/new.html.haml]: Yea, you have to learn how to modify it. its fine.

File: app/controllers/reset_passwords_controller.rb (hunk 4 above)
Comment (gvszg) [inline app/controllers/reset_passwords_controller.rb]: Thanks, got it.

**Pull request merged**